### PR TITLE
Revision of Quickpick for Theia

### PR DESCRIPTION
**Pull request by Colin-Stone**
Signed-off-by: Colin-Stone <30794003+Colin-Stone@users.noreply.github.com>
Using specific QuickPick functions rather than "Helper" functions to allow user to start typing straight away and that value to be interpreted. Makes the overall experience less of a surprise. However using these functions don't work with Theia so had to create a Theia specific alternative code path.
Also fixes: https://github.com/zowe/vscode-extension-for-zowe/issues/277 
and fixes: https://github.com/zowe/vscode-extension-for-zowe/issues/271

In a number of places  had a to add 
// tslint:disable-next-line: no-implicit-dependencies
import { Session, IProfileLoaded, Logger } from "@brightside/imperative";
Looks as if something changed with linter.

### CODE REVIEW

File: src/PersistentFilters.ts [hunk 1]
```diff
@@ -61,19 +61,21 @@ export class PersistentFilters {
...
-            return element.trim() !== criteria.trim();
-        });
-
-        // Add value to front of stack
-        this.mHistory.unshift(criteria);
-
-        // If list getting too large remove last entry
-        if (this.mHistory.length > this.maxHistory) {
-            this.mHistory.pop();
+        if (criteria) {
```
Comment (zFernand0) [inline src/PersistentFilters.ts]: :+1:

File: src/USSTree.ts [hunk 2]
```diff
@@ -276,33 +274,54 @@ export class USSTree implements vscode.TreeDataProvider<ZoweUSSNode> {
...
                 const options: vscode.InputBoxOptions = {
                     prompt: localize("ussFilterPrompt.option.prompt.search",
-                        "Search Unix System Services (USS) by entering a path name starting with a /"),
+                        "Create a new filter"),
                     value: sessionNode.fullPath
                 };
                 // get user input
                 remotepath = await vscode.window.showInputBox(options);
-                if (!remotepath) {
+                if (!remotepath || remotepath.length === 0) {
```
Comment (zFernand0) [inline src/USSTree.ts]: If I remember correctly empty strings evaluate to `falsy`, so don't know if the second part is even reachable.
No change required!

File: tslint.json [hunk 3]
```diff
@@ -37,8 +37,9 @@
       "check-module"
     ],
     "no-implicit-dependencies": [
-      true,
-      "dev"
+      false,
+      "dev",
+      ["@brightside/imperative"]
```
Comment (zFernand0) [inline tslint.json]: This will be change to `true, dev, imperative` by #288 
Sorry for the confusion
No change required!

File: src/ZoweJobNode.ts [hunk 4]
```diff
@@ -140,11 +119,6 @@ export class Job extends vscode.TreeItem {
         this.dirty = false;
         return this.children;
     }
-    public getDetailLabel(): string {
-        return this.contextValue === "Job" || this.contextValue === extension.JOBS_JOB_CONTEXT + extension.FAV_SUFFIX ?
-            `${this.job.jobname}(${this.job.jobid})`
-            : Job.createSearchLabel(this.owner, this.prefix, this.searchId);
-    }
```
Comment (zFernand0) [inline src/ZoweJobNode.ts]: I can see `createSearchLabel` being moved under `ZosJobsProvider.ts`, but maybe I missed this function. (Or maybe it isn't used anymore?

File: src/ZosJobsProvider.ts [hunk 5]
```diff
@@ -16,7 +16,7 @@ import { IProfileLoaded, Logger } from "@brightside/imperative";
 import { Profiles } from "./Profiles";
 import { PersistentFilters } from "./PersistentFilters";
 import { Job } from "./ZoweJobNode";
-import * as utils from "./utils";
+import { OwnerFilterDescriptor, JobIdFilterDescriptor, applyIcons, FilterItem, FilterDescriptor, resolveQuickPickHelper } from "./utils";
```
Comment (zFernand0) [inline src/ZosJobsProvider.ts]: :+1:

File: src/ZosJobsProvider.ts [hunk 6]
```diff
@@ -35,12 +35,12 @@ export async function createJobsTree(log: Logger) {
     return tree;
 }
 
+// tslint:disable-next-line: max-classes-per-file
```
Comment (zFernand0) [inline src/ZosJobsProvider.ts]: We should try to avoid disabling tslint rules as much as possible.
Is this something we should discuss further, and maybe modify `tslint.json`?

File: src/ZosJobsProvider.ts [hunk 7]
```diff
@@ -280,75 +282,114 @@ export class ZosJobsProvider implements vscode.TreeDataProvider<Job> {
...
-                    return;
+                if (!hasHistory || choice === this.createOwner) { // User has selected owner/prefix option
+                    options = {
+                        prompt: localize("jobsFilterPrompt.option.prompt.owner",
+                        "Enter the Job Owner. Default is *."),
+                        value: node.owner
+                    };
+                    // get user input
+                    owner = await vscode.window.showInputBox(options);
+                    if (owner === undefined) {
```
Comment (zFernand0) [inline src/ZosJobsProvider.ts]: I assume this is how VSCode APIs report a cancelled/escaped action. otherwise it will be empty or null?

Comment (Colin-Stone) [reply src/ZosJobsProvider.ts]: Exactly. If I checked for !owner I would be including if the user specific an empty string on purpose

File: __tests__/__unit__/USSTree.unit.test.ts [hunk 8]
```diff
@@ -353,7 +355,9 @@ describe("Unit Tests (Jest)", () => {
     /*************************************************************************************************************
      * USS Filter prompts
      *************************************************************************************************************/
-    it("Testing that user filter prompts are executed successfully", async () => {
+    it("Testing that user filter prompts are executed successfully, theia specific route", async () => {
+        let theia = true;
+        Object.defineProperty(extension, "ISTHEIA", { get: () => theia });
```
[the remark this answers is not in the document]
Comment (Colin-Stone) [reply __tests__/__unit__/USSTree.unit.test.ts]: Good point I didn't try that one

[the remark this answers is not in the document]
Comment (Colin-Stone) [reply __tests__/__unit__/USSTree.unit.test.ts]: I'm afraid the lines above and the subsequent theia = false was straight from google. As it work first time I didn't modify at all

File: __tests__/__unit__/USSTree.unit.test.ts [hunk 9]
```diff
@@ -383,7 +389,109 @@ describe("Unit Tests (Jest)", () => {
         await testTree.ussFilterPrompt(testTree.mSessionNodes[1]);
         expect(showInformationMessage.mock.calls.length).toBe(1);
         expect(showInformationMessage.mock.calls[0][0]).toBe("No selection made.");
+        theia = false;
```
[the remark this answers is not in the document]
Comment (Colin-Stone) [reply __tests__/__unit__/USSTree.unit.test.ts]: This was to put it back into VSCode only mode. It may not have been necessary outside of this particular it section but I didn't circle back to try it.

File: __tests__/__unit__/DatasetTree.unit.test.ts [hunk 10]
```diff
@@ -418,6 +424,102 @@ describe("DatasetTree Unit Tests", () => {
...
+        );
+        let entered;
+
+        // Assert edge condition user cancels the input path box
+        createQuickPick.mockReturnValue({
+            placeholder: "Select a filter",
+            activeItems: [qpItem],
+            ignoreFocusOut: true,
+            items: [qpItem],
+            value: entered,
```
[the remark this answers is not in the document]
Comment (Colin-Stone) [reply __tests__/__unit__/DatasetTree.unit.test.ts]: That was an oversight on my part

File: src/USSTree.ts (hunk 2 above)
Comment (Colin-Stone) [reply src/USSTree.ts]: You are correct. Ironically I found that out that whilst creating these tests but obviously didn't apply it here

File: tslint.json (hunk 3 above)
Comment (Colin-Stone) [reply tslint.json]: We should double check prior to the next version to make sure we haven't overwritten the correct final implementation

Comment (zFernand0) [reply tslint.json]: as long as this PR goes in before #288, we should be fine.
But agreed, let's double check once everything is on master :)

File: src/ZosJobsProvider.ts (hunk 6 above)
Comment (Colin-Stone) [reply src/ZosJobsProvider.ts]: I do use the max-classes a lotso maybe I am doing something wrong by having so many

File: __tests__/__unit__/USSTree.unit.test.ts [hunk 11]
```diff
@@ -383,7 +389,109 @@ describe("Unit Tests (Jest)", () => {
...
+    it("Testing that user filter prompts are executed successfully, VSCode route", async () => {
+        testTree.initialize(Logger.getAppLogger());
+        let qpItem: vscode.QuickPickItem = new utils.FilterDescriptor("\uFF0B " + "Create a new filter");
+        expect(qpItem.description).toBeFalsy();
+        expect(qpItem.alwaysShow).toBe(true);
+
+        const resolveQuickPickHelper = jest.spyOn(utils, "resolveQuickPickHelper").mockImplementation(
+            () => Promise.resolve(qpItem)
+        );
+        let entered;
```
[the remark this answers is not in the document]
Comment (Colin-Stone) [reply __tests__/__unit__/USSTree.unit.test.ts]: Yes all three should be the same

File: src/ZoweJobNode.ts (hunk 4 above)
Comment (Colin-Stone) [reply src/ZoweJobNode.ts]: getDetailLabel isn't referenced. Actually part of a commented out block which I should have removed/cleaned up
